src/Lykke.Service.ExchangeConnector/Exchanges/Concrete/GDAX/GdaxOrderBooksHarvester.cs

```diff
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Log;
+using TradingBot.Communications;
+using TradingBot.Exchanges.Concrete.GDAX.RestClient;
+using TradingBot.Exchanges.Concrete.GDAX.RestClient.Entities;
+using TradingBot.Exchanges.Concrete.GDAX.WssClient;
+using TradingBot.Exchanges.Concrete.GDAX.WssClient.Entities;
+using TradingBot.Exchanges.Concrete.Shared;
+using TradingBot.Infrastructure.Configuration;
+
+namespace TradingBot.Exchanges.Concrete.GDAX
+{
+    internal sealed class GdaxOrderBooksHarvester : OrderBooksHarvesterBase
+    {
+        private readonly GdaxExchangeConfiguration _configuration;
+        private readonly GdaxWebSocketApi _websocketApi;
+        private readonly GdaxRestApi _restApi;
+
+        public GdaxOrderBooksHarvester(GdaxExchangeConfiguration configuration, ILog log,
+            OrderBookSnapshotsRepository orderBookSnapshotsRepository, OrderBookEventsRepository orderBookEventsRepository)
+            : base(configuration, log, orderBookSnapshotsRepository, orderBookEventsRepository)
+        {
+            _configuration = configuration;
+            _websocketApi = CreateWebSocketsApiClient();
+            _restApi = CreateRestApiClient();
+        }
+
+        protected override async Task MessageLoopImpl()
+        {
+            try
+            {
+                await _websocketApi.ConnectAsync(CancellationToken);
+                // TODO: First subscribe with websockets and ignore the events before the GetOpenOrders execution
+                await HandleOpenedOrders(await _restApi.GetOpenOrders(CancellationToken)); // Send symbol 
+
+                await _websocketApi.SubscribeToFullUpdatesAsync(
+                    _configuration.Instruments.Select(ConvertSymbolFromLykkeToExchange).ToArray(),
+                    CancellationToken);
+            }
+            finally
+            {
+                try
+                {
+                    using (var cts = new CancellationTokenSource())
```

Comment:
CancellationTokenSource has a constructor with a timeout, I think it is simpler to use it